README.md
# Conjur Azure DevOps Extension
Azure Devops Task Extension with API Key Authentication mechanism for supporting batch retrieval of secrets from the Cyberark Conjur Secrets Manager in secured way through Azure DevOps Pipeline.

The Authentication parameters to connect to Conjur Server are configured in Service Connection.

## Certification Level
![](https://img.shields.io/badge/Certification%20Level-Community-28A745?link=https://github.com/cyberark/community/blob/master/Conjur/conventions/certification-levels.md)

This repo is a **Community** level project. It's a community contributed project that **is not reviewed or supported
by CyberArk**. For more detailed information on our certification levels, see [our community guidelines](https://github.com/cyberark/community/blob/master/Conjur/conventions/certification-levels.md#community).

## Features

The following features are available with the Azure DevOps Extension:

* API authentication
* Batch retrieval of secrets from Conjur Server with help of secured Service Connection

## Limitations

The Azure DevOps Extension does not support creating, updating or removing secrets

## Technical Requirements

| Technology             | Version  |
| ---------------------- | -------- |
| Conjur OSS             | 1.9+     |
| Conjur Enterprise      | 12.5+    |
| Conjur Cloud           |          |
| Azure DevOps account   |          |

# Prerequisites

The following are prerequisites to use Azure DevOps extension.

## Conjur setup

Conjur (OSS or Enterprise or Cloud) and the Conjur CLI are installed in the environment and running in the background.

This section describes how to set up the API Authentication for Conjur OSS or Enterprise

1. Define the API Authentication policy

- Create a policy that defines the API Authentication, for example BotApp.
```yaml
- !policy
  id: BotApp
  body:
    # Define a human user, a non-human identity that represents an application, and a secret
  - !user Dave
  - !host myDemoApp
  - &variables
      - !variable secretVar
      - !variable some-other-secret
  - !permit
    # Give permissions to the human user to update the secret and fetch the secret.
    role: !user Dave
    privileges: [read, update, execute]
    resource: *variables
  - !permit
    # Give permissions to the non-human identity to fetch the secret.
    role: !host myDemoApp
    privileges: [read, execute]
    resource: *variables
```

- Save the policy as BotApp.yml, and load it to root:

```
     conjur policy load -b root -f /path/to/file/authn-iam.yml
```
     
- Conjur generates the following API keys:
     - An API key for Dave, the human user. This key is used to authenticate user Dave to Conjur.
     - An API key for BotApp, the non-human identity. This key is used to authenticate BotApp application to Conjur.

     Those API keys is correlated with the number of Users & Hosts defined in a policy.

2.  Set the secret    
     a. Generate a secret

     Generate a value for your application’s secret:
     ```
     secretVal=$(openssl rand -hex 12 | tr -d '\r\n')
     ```

     This generates a 12-hex-character value.

     b. Store the secret

     Store the generated value in Conjur:
     ```
     conjur variable set -i BotApp/secretVar -v ${secretVal}
     ```

     A policy predefined variable named `BotApp/secretVar` is set with a random
     generated secret.

This section describes how to set up the API Authentication for Conjur Cloud

Pre-requisite:
- Conjur cloud
- Pcloud
- Conjur cloud cli

<small><a href='https://docs-er.cyberark.com/ConjurCloud/en/Content/ConjurCloud/ccl-manage-users.htm?tocpath=Get%20started%7CTutorial%7C_____1'>Refer the tutorial for Conjur Cloud Setup </a></small>
1. Manage Conjur Cloud users
2. Set up the Conjur Cloud CLI
3. Log in to Conjur Cloud
4. Sync Privilege Cloud Safe to Conjur

To create a simple host that authenticates using an API key:

Create a policy for the host:

```yaml
- !host
  id: <host name>
  annotations:
    authn/api-key: true
```    
Save the policy as myapp-host.yaml.

Load the policy file into the data policy branch:

```
conjur policy load -b data -f myapp-host.yaml
```

To grant permissions on secrets:

```yaml
- !grant
  role: !group delegation/consumers
  member: !host /data/myapp
```

Save the file as grant_permissions.yml.

Load the policy to data/vault/secrets-safe:

```
conjur policy load -b data/vault/secrets-safe -f grant_permissions.yml
```

## Azure DevOps Setup

* Download Conjur Azure DevOps Extension from Azure Marketplace
* Install the extension to an Azure DevOps organization
* Search for installed extension in Project Settings > Pipelines > Service connection > Create service connection

     <img src="https://github.com/cyberark/conjur-azure-devops-extension/blob/main/images/service-connection.png" width="300" height="300">

* Add the Conjur details in Service Connection 

     <img src="https://github.com/cyberark/conjur-azure-devops-extension/blob/main/images/setupSC.png" width="400" height="500">

* In Pipeline > Task > Search with Batch Secret Retrieval > Select the Service Connection and provide path of secrets.yml file

     <img src="https://github.com/cyberark/conjur-azure-devops-extension/blob/main/images/pipelineTask.png" width="500" height="500">

* secrets.yml file format  

- Conjur OSS or Enterprise

```yaml
SECRET: !var BotApp/secretVar
ANOTHER_SECRET: !var some-other-secret
```

- Conjur Cloud

```yaml
SECRET: !var data/vault/secrets-safe/ado_secret_apikey/address
ANOTHER_SECRET: !var data/vault/secrets-safe/ado_secret_apikey/username
```

* Under steps in azure-pipeline.yml task is added

```yaml
steps:
- task: secretBatchRetrievalConnector@0
  displayName: ConjurIntegeration
  inputs:
    ConjurService: 'ConjurSConnection'
    secretsyml: './secrets.yml'
```

## Development

Please follow this guide to properaly set up this extension:
https://docs.microsoft.com/en-us/azure/devops/extend/develop/add-build-task?view=azure-devops

## Contributing
We welcome contributions of all kinds to this repository. For instructions on how to get started and descriptions
of our development workflows, please see our [contributing guide](CONTRIBUTING.md).

## License
This repository is licensed under Apache License 2.0 - see [`LICENSE`](LICENSE) for more details.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added
## [1.0.0]
- Extension now supports to Fetch Bulk secrets from Conjur Vault

secretBatchRetrievalConnector/test/_suite.js
"use strict";
exports.__esModule = true;
var path = require("path");
var assert = require("assert");
var ttm = require("azure-pipelines-task-lib/mock-test");
describe("Fetch secret", function () {
    before(function () {
    });
    after(function () {
    });
    it('should succeed with valid inputs', function (done) {
        this.timeout(10000);
        var tp = path.join(__dirname, 'success.js');
        var tr = new ttm.MockTestRunner(tp);
        tr.run();
        console.log(tr.stdout);
        assert.ok(tr.stdout.indexOf('SECRET_WITH_SLASHES') >= 0, "should display the secret");
        done();
    });
    it('should fail with incorrect endpoint input', function (done) {
        this.timeout(10000);
        var tp = path.join(__dirname, 'failure.js');
        var tr = new ttm.MockTestRunner(tp);
        tr.run();
        console.log(tr.stdout);
        console.log("printv:: ", tr.errorIssues);
        assert.ok(tr.stdout.indexOf('SECRET_WITH_SLASHES') == -1, "should not display the secret");
        done();
    });
    it('should fail with incorrect secret variables', function (done) {
        this.timeout(10000);
        var tp = path.join(__dirname, 'invalid_var.js');
        var tr = new ttm.MockTestRunner(tp);
        tr.run();
        console.log(tr.stdout);
        console.log("printv:: ", tr.errorIssues);
        assert.ok(tr.errorIssues.length > 0, "should not display the secret as invalid secret variables");
        done();
    });
});


secretBatchRetrievalConnector/test/_suite.ts
import * as path from 'path';
import * as assert from 'assert';
import * as ttm from 'azure-pipelines-task-lib/mock-test';

describe("Fetch secret" , () => {
    before(() =>{

    });

    after(() =>{

    });

    it('should succeed with valid inputs', function(done: Mocha.Done) {
        this.timeout(10000);

        let tp = path.join(__dirname, 'success.js');
        let tr: ttm.MockTestRunner = new ttm.MockTestRunner(tp);

        tr.run();
        console.log(tr.stdout);
        assert.ok(tr.stdout.indexOf('SECRET_WITH_SLASHES') >= 0, "should display the secret");
        done();
    });

    it('should fail with incorrect endpoint input', function(done: Mocha.Done) {
        this.timeout(10000);

        let tp = path.join(__dirname, 'failure.js');
        let tr: ttm.MockTestRunner = new ttm.MockTestRunner(tp);

        tr.run();
        console.log(tr.stdout);
        console.log("printv:: ",tr.errorIssues);
        assert.ok(tr.stdout.indexOf('SECRET_WITH_SLASHES') == -1, "should not display the secret");
        done();
    });  

    it('should fail with incorrect secret variables', function(done: Mocha.Done) {
        this.timeout(10000);

        let tp = path.join(__dirname, 'invalid_var.js');
        let tr: ttm.MockTestRunner = new ttm.MockTestRunner(tp);

        tr.run();
        console.log(tr.stdout);
        console.log("printv:: ",tr.errorIssues);
        assert.ok(tr.errorIssues.length > 0, "should not display the secret as invalid secret variables");
        done();
    });
})

secretBatchRetrievalConnector/test/failure.js
"use strict";
exports.__esModule = true;
var tmrm = require("azure-pipelines-task-lib/mock-run");
var path = require("path");
var taskPath = path.join(__dirname, '..', 'index.js');
var tmr = new tmrm.TaskMockRunner(taskPath);
var endpointId = "ConjurService";
tmr.setInput(endpointId, endpointId);
process.env['ENDPOINT_URL_' + endpointId] = "http://localhost";
process.env["ENDPOINT_DATA_" + endpointId + "_" + 'conjuraccount'.toUpperCase()] = 'myConjurAccount';
process.env["ENDPOINT_DATA_" + endpointId + "_" + 'conjurusername'.toUpperCase()] = 'host/BotApp/myDemoApp';
process.env["ENDPOINT_DATA_" + endpointId + "_" + 'conjurapikey'.toUpperCase()] = '3ded3pp1x1w3ak3ter2kj22532mx2xfk51z2p28p';
tmr.setInput('secretsyml', path.join(__dirname, '..', 'test/test-secrets.yml'));
tmr.run();


secretBatchRetrievalConnector/index.ts
import tl = require('azure-pipelines-task-lib/task');
import https = require('https');
import * as fs from 'fs';
const readline = require('readline');
const file_type_supported: string[] = ['.yml', '.yaml'];

// utility functions
function trimRight(input : string, trimStr : string) : string{
    if (input?.endsWith(trimStr)) {
        input = input.substring(0, input.length-1)
        return trimRight(input, trimStr);
    }
    return input;
}

function removeHttps(input : string) : string {
    return input.replace("https://", "").replace("http://", "")
}

function base64(input : string) : string {
    return  Buffer.from(input).toString('base64')
}

function removeApi(input : string){ 
    if (input.endsWith('/api')) { 
       return [input.replace("/api", ""), "/api"]
    } else {
        return [input, ""]
    }    
}

function getTokenHeader(token : string) {
    return "Token token=\"" + base64(token) + "\"";
}

function sendHttpRequest(hostname : string, endpoint : string, method : 'GET' | 'POST', authorization : string, data : string, ignoreSsl : boolean) {
    // very helpful for debugging but does leak passwords/tokens when in debug mode
    // console.debug(`------------\n${method} ${hostname}${endpoint}\nAuthorization: ${authorization}\n\n${data}`)
    
    if(ignoreSsl) {
        // this will auto prompt with warning
        process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";
    }

    // posting data must include Content-Length in HTTP header
    var dataLength = 0;
    if (data) {
        dataLength = data.length;
    }

    return new Promise((resolve, reject) => {     
        const options = {
            hostname: hostname,
            port: 443,
            path: endpoint,
            method: method,
            headers: {
                'Content-Type': 'application/json',
                'Content-Length': dataLength,
                'Authorization': authorization
            }
        }

        var responseBody = []
        
        const req = https.request(options, (res) => {
            res.on('data', (chunk) => responseBody.push(chunk));
            res.on('end', () => resolve(responseBody.join('')));

            // If non-200 code is returned from any call using this method the task extension will fail
            if(res.statusCode != 200){
                tl.setResult(tl.TaskResult.Failed, `recieved status code '${res.statusCode}': ${responseBody.join('')}`);
            }
        })
        
        req.on('error', (error) => {
            tl.setResult(tl.TaskResult.Failed, `${error}`);
        })
        if(data) {
            req.write(data)
        }
        req.end()
    })
}

// conjur api functions
function authenticate(hostname: string, account : string, username : string, apiKey : string, type : AuthnTypes, ignoreSsl : boolean) {
    var [hname, api] = removeApi(hostname);
    switch(type) {
        case AuthnTypes.ApiKey:
            username = encodeURIComponent(username);
            var endpoint = `${api}/authn/${account}/${username}/authenticate`;
            return sendHttpRequest(hname, endpoint, 'POST', "", apiKey, ignoreSsl);
        default:
            tl.setResult(tl.TaskResult.Failed, `Invalid authentication type '${type}'. Valid types are 'apiKey'`)
    }
}

function getSecretsPath(hostname: string, token : string, ignoreSsl : boolean, secretPath: any[]){
    var [hname, api] = removeApi(hostname);
    var endpoint = `${api}/secrets?variable_ids=${secretPath}`;
    token = getTokenHeader(token);
    return sendHttpRequest(hname, endpoint, 'GET', token, null, ignoreSsl)
}

function setAzureSecrets(jsonData: string, secretPaths: ISecret, hostname: string, account : string, token : string, debug_mode: boolean, ignoreSsl : boolean){
    var conjurSecret = JSON.parse(jsonData);
    if ('error' in conjurSecret){
        console.log("Batch retrieval failed, falling back to single secret fetch at a time");
        for (let ele in secretPaths){
            getASecret(hostname, account, token, ele, ignoreSsl)
                .then((data) => {
                    tl.setVariable(secretPaths[ele], data.toString(), true);
                    if(debug_mode == true)
                      console.log(`Set conjur secret '${ele}' to azure variable '${secretPaths[ele]}'`);
                })
                .catch((err) => tl.setResult(tl.TaskResult.Failed, err.message)
            )
        }
    }
    else{
        console.log("Batch retrieval successful");
        for (let key in conjurSecret) {
            var items: string[] = key.split(":");
            var ele: string = items[items.length-1];
            tl.setVariable(secretPaths[ele], conjurSecret[key], true);
            if(debug_mode == true)
               console.log(`Set conjur secret '${ele}' to azure variable '${secretPaths[ele]}'`);
        }
    }
}

function getASecret(hostname: string, account : string, token : string, secretId: string, ignoreSsl : boolean){
    var [hname, api] = removeApi(hostname);
    secretId = encodeURIComponent(secretId);
    var endpoint = `${api}/secrets/${account}/variable/${secretId}`;
    token = getTokenHeader(token);
    return sendHttpRequest(hname, endpoint, 'GET', token, null, ignoreSsl);
}

function batchSecretRetrieval(hostname: string, account : string, token : string, secretYml : string,  debug_mode: boolean, ignoreSsl : boolean){
    var secretsPath = [];
    var secret:ISecret = {}
    
    const readInterface = readline.createInterface({
        input: fs.createReadStream(secretYml),
        output: process.stdout,
        console: false
    });
    readInterface.on('line', function(line) {
        if (line.toString().includes(': !var')) {
            createISecret(line, secret);       
        }else{
            tl.setResult(tl.TaskResult.Failed, `Secret path not found in required format at line '${line}'. Format required eg: "SECRET_VAR: !var /secret/path"`);  
        }
    }).on('close', function(line) {
        for (let key in secret) {
            secretsPath.push(account + ":variable:" + encodeURIComponent(key));
        }
        if (secretsPath.length > 0){
            getSecretsPath(hostname, token, ignoreSsl, secretsPath)
            .then((jsonData) => setAzureSecrets(jsonData.toString(), secret, hostname, account, token, debug_mode, ignoreSsl))
            .catch((err) => tl.setResult(tl.TaskResult.Failed, err.message))
        }else{
            return tl.setResult(tl.TaskResult.Failed, 'No data or secret path not found in required format'); 
        }    
    }).on('error', function(err) {
        tl.setResult(tl.TaskResult.Failed, err.message);
    });
}

// task extension functions
type ISecret = { 
    [key: string]: string
} 

function createISecret(line : string, secret: ISecret):ISecret {
    var secretSections = line.split(": !var ");
    if(secretSections.length != 2) {
        tl.setResult(tl.TaskResult.Failed, `Failed to retrieve secret name and path from '${line}'`)
    }
    var secretName = secretSections[0].trim();
    var secretPath = secretSections[1].trim();
    secret[secretPath] = secretName;
    return secret;
}

enum AuthnTypes {
    ApiKey,
    AzureManagedIdentity
}

async function run() {
    try {
        //input from service connector
        var ep = tl.getInput('ConjurService', true);
        var debug_mode: boolean; 
        var ignoreSsl: boolean; 
        var hostname = tl.getEndpointUrlRequired(ep);
        var account  = tl.getEndpointDataParameter(ep, 'conjuraccount', true);
        var username = tl.getEndpointDataParameter(ep, 'conjurusername', true);
        var apiKey = tl.getEndpointDataParameter(ep, 'conjurapikey', true);
        if(tl.getVariable('System.Debug') == 'true')
            debug_mode = true;
        else
            debug_mode = false;

        if(tl.getEndpointDataParameter(ep, 'ignoressl', true) == 'true')
           ignoreSsl = true;
        else
           ignoreSsl = false;   
    
        var secretYml = tl.getInput('secretsyml', false);
        var clientId = tl.getInput('azureclientid', false);    
        var authnType = AuthnTypes.ApiKey

        // Set defaults
        if (!hostname) {
            hostname = "";
        }
        if (!account) {
            account = "";
        }
        if (!username) {
            username = "";
        }
        if (!apiKey) {
            apiKey = "";
        }
        if (!secretYml) {
            secretYml = "./secrets.yml";
        }else{
            var path = require('path');
            if (!file_type_supported.includes(path.extname(secretYml).toLowerCase())){
                return tl.setResult(tl.TaskResult.Failed, "Only .yml, .yaml extension file supported");
            } 
        }
        if (!clientId) {
            clientId = "";
            // client id is not provided assume api key authentication
            if (apiKey == "") {
                var error_message = "Conjur API Key is required since Azure Client ID is not provided";
                console.error(error_message);
                tl.setResult(tl.TaskResult.Failed, error_message);
            }
        } else {
            // If client id is provided we are default to azure authentication
            authnType = AuthnTypes.AzureManagedIdentity
            console.log("Since Client ID was provided then defaulting to azure authentication. Api key will not be used")
        }

        // sanitize
        hostname = trimRight(hostname, '/');
        hostname = removeHttps(hostname);

        // fetch the secrets
        authenticate(hostname, account, username, apiKey, authnType, true)
        .then((data) => batchSecretRetrieval(hostname, account, data.toString(), secretYml, debug_mode, ignoreSsl))
        .catch((err) => tl.setResult(tl.TaskResult.Failed, err.message))
    }
    catch (err) {
        tl.setResult(tl.TaskResult.Failed, err.message);
    }
}

run();

secretBatchRetrievalConnector/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
var tl = require("azure-pipelines-task-lib/task");
var https = require("https");
var fs = require("fs");
var readline = require('readline');
var file_type_supported = ['.yml', '.yaml'];
// utility functions
function trimRight(input, trimStr) {
    if (input === null || input === void 0 ? void 0 : input.endsWith(trimStr)) {
        input = input.substring(0, input.length - 1);
        return trimRight(input, trimStr);
    }
    return input;
}
function removeHttps(input) {
    return input.replace("https://", "").replace("http://", "");
}
function base64(input) {
    return Buffer.from(input).toString('base64');
}
function removeApi(input) {
    if (input.endsWith('/api')) {
        return [input.replace("/api", ""), "/api"];
    }
    else {
        return [input, ""];
    }
}
function getTokenHeader(token) {
    return "Token token=\"" + base64(token) + "\"";
}
function sendHttpRequest(hostname, endpoint, method, authorization, data, ignoreSsl) {
    // very helpful for debugging but does leak passwords/tokens when in debug mode
    // console.debug(`------------\n${method} ${hostname}${endpoint}\nAuthorization: ${authorization}\n\n${data}`)
    if (ignoreSsl) {
        // this will auto prompt with warning
        process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";
    }
    // posting data must include Content-Length in HTTP header
    var dataLength = 0;
    if (data) {
        dataLength = data.length;
    }
    return new Promise(function (resolve, reject) {
        var options = {
            hostname: hostname,
            port: 443,
            path: endpoint,
            method: method,
            headers: {
                'Content-Type': 'application/json',
                'Content-Length': dataLength,
                'Authorization': authorization
            }
        };
        var responseBody = [];
        var req = https.request(options, function (res) {
            res.on('data', function (chunk) { return responseBody.push(chunk); });
            res.on('end', function () { return resolve(responseBody.join('')); });
            // If non-200 code is returned from any call using this method the task extension will fail
            if (res.statusCode != 200) {
                tl.setResult(tl.TaskResult.Failed, "recieved status code '" + res.statusCode + "': " + responseBody.join(''));
            }
        });
        req.on('error', function (error) {
            tl.setResult(tl.TaskResult.Failed, "" + error);
        });
        if (data) {
            req.write(data);
        }
        req.end();
    });
}
// conjur api functions
function authenticate(hostname, account, username, apiKey, type, ignoreSsl) {
    var _a = removeApi(hostname), hname = _a[0], api = _a[1];
    switch (type) {
        case AuthnTypes.ApiKey:
            username = encodeURIComponent(username);
            var endpoint = api + "/authn/" + account + "/" + username + "/authenticate";
            return sendHttpRequest(hname, endpoint, 'POST', "", apiKey, ignoreSsl);
        default:
            tl.setResult(tl.TaskResult.Failed, "Invalid authentication type '" + type + "'. Valid types are 'apiKey'");
    }
}
function getSecretsPath(hostname, token, ignoreSsl, secretPath) {
    var _a = removeApi(hostname), hname = _a[0], api = _a[1];
    var endpoint = api + "/secrets?variable_ids=" + secretPath;
    token = getTokenHeader(token);
    return sendHttpRequest(hname, endpoint, 'GET', token, null, ignoreSsl);
}
function setAzureSecrets(jsonData, secretPaths, hostname, account, token, debug_mode, ignoreSsl) {
    var conjurSecret = JSON.parse(jsonData);
    if ('error' in conjurSecret) {
        console.log("Batch retrieval failed, falling back to single secret fetch at a time");
        var _loop_1 = function (ele_1) {
            getASecret(hostname, account, token, ele_1, ignoreSsl)
                .then(function (data) {
                tl.setVariable(secretPaths[ele_1], data.toString(), true);
                if (debug_mode == true)
                    console.log("Set conjur secret '" + ele_1 + "' to azure variable '" + secretPaths[ele_1] + "'");
            })["catch"](function (err) { return tl.setResult(tl.TaskResult.Failed, err.message); });
        };
        for (var ele_1 in secretPaths) {
            _loop_1(ele_1);
        }
    }
    else {
        console.log("Batch retrieval successful");
        for (var key in conjurSecret) {
            var items = key.split(":");
            var ele = items[items.length - 1];
            tl.setVariable(secretPaths[ele], conjurSecret[key], true);
            if (debug_mode == true)
                console.log("Set conjur secret '" + ele + "' to azure variable '" + secretPaths[ele] + "'");
        }
    }
}
function getASecret(hostname, account, token, secretId, ignoreSsl) {
    var _a = removeApi(hostname), hname = _a[0], api = _a[1];
    secretId = encodeURIComponent(secretId);
    var endpoint = api + "/secrets/" + account + "/variable/" + secretId;
    token = getTokenHeader(token);
    return sendHttpRequest(hname, endpoint, 'GET', token, null, ignoreSsl);
}
function batchSecretRetrieval(hostname, account, token, secretYml, debug_mode, ignoreSsl) {
    var secretsPath = [];
    var secret = {};
    var readInterface = readline.createInterface({
        input: fs.createReadStream(secretYml),
        output: process.stdout,
        console: false
    });
    readInterface.on('line', function (line) {
        if (line.toString().includes(': !var')) {
            createISecret(line, secret);
        }
        else {
            tl.setResult(tl.TaskResult.Failed, "Secret path not found in required format at line '" + line + "'. Format required eg: \"SECRET_VAR: !var /secret/path\"");
        }
    }).on('close', function (line) {
        for (var key in secret) {
            secretsPath.push(account + ":variable:" + encodeURIComponent(key));
        }
        if (secretsPath.length > 0) {
            getSecretsPath(hostname, token, ignoreSsl, secretsPath)
                .then(function (jsonData) { return setAzureSecrets(jsonData.toString(), secret, hostname, account, token, debug_mode, ignoreSsl); })["catch"](function (err) { return tl.setResult(tl.TaskResult.Failed, err.message); });
        }
        else {
            return tl.setResult(tl.TaskResult.Failed, 'No data or secret path not found in required format');
        }
    }).on('error', function (err) {
        tl.setResult(tl.TaskResult.Failed, err.message);
    });
}
function createISecret(line, secret) {
    var secretSections = line.split(": !var ");
    if (secretSections.length != 2) {
        tl.setResult(tl.TaskResult.Failed, "Failed to retrieve secret name and path from '" + line + "'");
    }
    var secretName = secretSections[0].trim();
    var secretPath = secretSections[1].trim();
    secret[secretPath] = secretName;
    return secret;
}
var AuthnTypes;
(function (AuthnTypes) {
    AuthnTypes[AuthnTypes["ApiKey"] = 0] = "ApiKey";
    AuthnTypes[AuthnTypes["AzureManagedIdentity"] = 1] = "AzureManagedIdentity";
})(AuthnTypes || (AuthnTypes = {}));
function run() {
    return __awaiter(this, void 0, void 0, function () {
        var ep, debug_mode, ignoreSsl, hostname, account, username, apiKey, secretYml, clientId, authnType, path, error_message;
        return __generator(this, function (_a) {
            try {
                ep = tl.getInput('ConjurService', true);
                hostname = tl.getEndpointUrlRequired(ep);
                account = tl.getEndpointDataParameter(ep, 'conjuraccount', true);
                username = tl.getEndpointDataParameter(ep, 'conjurusername', true);
                apiKey = tl.getEndpointDataParameter(ep, 'conjurapikey', true);
                if (tl.getVariable('System.Debug') == 'true')
                    debug_mode = true;
                else
                    debug_mode = false;
                if (tl.getEndpointDataParameter(ep, 'ignoressl', true) == 'true')
                    ignoreSsl = true;
                else
                    ignoreSsl = false;
                secretYml = tl.getInput('secretsyml', false);
                clientId = tl.getInput('azureclientid', false);
                authnType = AuthnTypes.ApiKey;
                // Set defaults
                if (!hostname) {
                    hostname = "";
                }
                if (!account) {
                    account = "";
                }
                if (!username) {
                    username = "";
                }
                if (!apiKey) {
                    apiKey = "";
                }
                if (!secretYml) {
                    secretYml = "./secrets.yml";
                }
                else {
                    path = require('path');
                    if (!file_type_supported.includes(path.extname(secretYml).toLowerCase())) {
                        return [2 /*return*/, tl.setResult(tl.TaskResult.Failed, "Only .yml, .yaml extension file supported")];
                    }
                }
                if (!clientId) {
                    clientId = "";
                    // client id is not provided assume api key authentication
                    if (apiKey == "") {
                        error_message = "Conjur API Key is required since Azure Client ID is not provided";
                        console.error(error_message);
                        tl.setResult(tl.TaskResult.Failed, error_message);
                    }
                }
                else {
                    // If client id is provided we are default to azure authentication
                    authnType = AuthnTypes.AzureManagedIdentity;
                    console.log("Since Client ID was provided then defaulting to azure authentication. Api key will not be used");
                }
                // sanitize
                hostname = trimRight(hostname, '/');
                hostname = removeHttps(hostname);
                // fetch the secrets
                authenticate(hostname, account, username, apiKey, authnType, true)
                    .then(function (data) { return batchSecretRetrieval(hostname, account, data.toString(), secretYml, debug_mode, ignoreSsl); })["catch"](function (err) { return tl.setResult(tl.TaskResult.Failed, err.message); });
            }
            catch (err) {
                tl.setResult(tl.TaskResult.Failed, err.message);
            }
            return [2 /*return*/];
        });
    });
}
run();
